File: src/android/com.clk.promises/src/com/clk/promises/PromiseBase.java
package com.clk.promises;


import java.util.ArrayList;
import com.clk.Action;


public class PromiseBase<TResult> {

	// enumerations
    private enum PromiseState {
        Pending,  // unresolved  
        Resolved, // has-resolution
        Rejected, // has-rejection
    };
    
    protected enum ResultType {
    	Empty,
    	EmptyPromise,
        New,
        NewPromise,
    }
    
    
	// fields
    private Object _syncRoot = new Object();

    private PromiseState _state = PromiseState.Pending;

    private TResult _result = null;

    private Exception _error = null;    
    
    private ArrayList<Action.Type1<TResult>> _resolveHandlers = null;

    private ArrayList<Action.Type1<Exception>> _rejectHandlers = null;

    private ArrayList<Action.Type1<Progress>> _notifyHandlers = null;

    private ArrayList<Action.Type1<Progress>> _notifyHandlersSnapshot = null;
        
    
    // constructors
    protected PromiseBase() { }
 	
    
    // methods 
    protected void push(Action.Type1<TResult> resolveHandler, Action.Type1<Exception> rejectHandler, Action.Type1<Progress> notifyHandler)
    {
        // contracts
        if (resolveHandler == null) throw new IllegalArgumentException();
        if (rejectHandler == null) throw new IllegalArgumentException();
        if (notifyHandler == null) throw new IllegalArgumentException();

        // sync         
        synchronized (_syncRoot)
        {
            // resolveHandler
            if (_state == PromiseState.Pending)
            {
                if(_resolveHandlers==null)
                {
                    _resolveHandlers = new ArrayList<Action.Type1<TResult>>();
                }
                _resolveHandlers.add(resolveHandler);
            }

            // rejectHandler
            if (_state == PromiseState.Pending)
            {
                if (_rejectHandlers == null)
                {
                    _rejectHandlers = new ArrayList<Action.Type1<Exception>>();
                }
                _rejectHandlers.add(rejectHandler);
            }

            // notifyHandler
            if (_notifyHandlers == null)
            {
                _notifyHandlers = new ArrayList<Action.Type1<Progress>>();
            }
            _notifyHandlers.add(notifyHandler);
            _notifyHandlersSnapshot = null;

            // pending
            if (_state == PromiseState.Pending)
            {
                return;
            }
        }

        // resolved
        if (_state == PromiseState.Resolved)
        {
            try {
				resolveHandler.raise(_result);
			} catch (Exception ex) {
				throw new IllegalStateException(ex.getMessage(), ex.getCause());
			}
        }

        // rejected
        if (_state == PromiseState.Rejected)
        {
            try {
				rejectHandler.raise(_error);
            } catch (Exception ex) {
				throw new IllegalStateException(ex.getMessage(), ex.getCause());
			}
        }
    }    
            
    protected void resolveBase(TResult result)
    {
        // sync
        synchronized (_syncRoot)
        {
            // state
            if (_state != PromiseState.Pending) return;
            _state = PromiseState.Resolved;

            // results
            _result = result;
            _error = null;
        }

        // handlers
        ArrayList<Action.Type1<TResult>> resolveHandlers = _resolveHandlers;
        if (resolveHandlers == null) return;

        // resolve
        for (Action.Type1<TResult> resolveHandler : resolveHandlers)
        {
        	try {
				resolveHandler.raise(result);
        	} catch (Exception ex) {
				throw new IllegalStateException(ex.getMessage(), ex.getCause());
			}
        }
    }

    public void reject(Exception error)
    {
        // contracts
        if (error == null) throw new IllegalArgumentException();

        // sync            
        synchronized (_syncRoot)
        {
            // state
            if (_state != PromiseState.Pending) return;
            _state = PromiseState.Rejected;

            // result
            _result = null;
            _error = error;
        }

        // handlers
        ArrayList<Action.Type1<Exception>> rejectHandlers = _rejectHandlers;
        if (rejectHandlers == null) return;

        // reject
        for (Action.Type1<Exception> rejectHandler : rejectHandlers)
        {
        	try {
				rejectHandler.raise(error);
        	} catch (Exception ex) {
				throw new IllegalStateException(ex.getMessage(), ex.getCause());
			}
        }
    }

    public void notify(Progress progress)
    {
    	// contracts
        if (progress == null) throw new IllegalArgumentException();

        // handlers
        ArrayList<Action.Type1<Progress>> notifyHandlers = null;            
        synchronized (_syncRoot)
        {
            if (_notifyHandlersSnapshot == null && _notifyHandlers != null)
            {
                _notifyHandlersSnapshot = new ArrayList<Action.Type1<Progress>>();
                for (Action.Type1<Progress> notifyHandler : _notifyHandlers)
                {
                	_notifyHandlersSnapshot.add(notifyHandler);
                }                
            }
            notifyHandlers = _notifyHandlersSnapshot;
        }
        if (notifyHandlers == null) return;

        // notify
        for (Action.Type1<Progress> notifyHandler : notifyHandlers)
        {
        	try {
				notifyHandler.raise(progress);
        	} catch (Exception ex) {
				throw new IllegalStateException(ex.getMessage(), ex.getCause());
			}
        }
    }
}

File: src/windows/CLK.Promises/PromiseBase.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CLK.Promises
{
    public abstract class PromiseBase<TResult>
    {
        // Enumerations
        private enum PromiseState
        {
            Pending,  // unresolved  
            Resolved, // has-resolution
            Rejected, // has-rejection
        };

        protected enum ResultType
        {
            Empty,
            EmptyPromise,
            New,
            NewPromise,
        }


        // Fields
        private Object _syncRoot = new Object();

        private PromiseState _state = PromiseState.Pending;

        private TResult _result = default(TResult);

        private Exception _error = null;

        private List<Action<TResult>> _resolveHandlers = null;

        private List<Action<Exception>> _rejectHandlers = null;

        private List<Action<Progress>> _notifyHandlers = null;

        private List<Action<Progress>> _notifyHandlersSnapshot = null;


        // Methods 
        protected void Push(Action<TResult> resolveHandler, Action<Exception> rejectHandler, Action<Progress> notifyHandler)
        {
            #region Contracts

            if (resolveHandler == null) throw new ArgumentNullException();
            if (rejectHandler == null) throw new ArgumentNullException();
            if (notifyHandler == null) throw new ArgumentNullException();

            #endregion

            // Sync         
            lock (_syncRoot)
            {
                // ResolveHandler
                if (_state == PromiseState.Pending)
                {
                    if (_resolveHandlers == null)
                    {
                        _resolveHandlers = new List<Action<TResult>>();
                    }
                    _resolveHandlers.Add(resolveHandler);
                }

                // RejectHandler
                if (_state == PromiseState.Pending)
                {
                    if (_rejectHandlers == null)
                    {
                        _rejectHandlers = new List<Action<Exception>>();
                    }
                    _rejectHandlers.Add(rejectHandler);
                }

                // NotifyHandler
                if (_notifyHandlers == null)
                {
                    _notifyHandlers = new List<Action<Progress>>();
                }
                _notifyHandlers.Add(notifyHandler);
                _notifyHandlersSnapshot = null;

                // Pending
                if (_state == PromiseState.Pending)
                {
                    return;
                }
            }

            // Resolved
            if (_state == PromiseState.Resolved)
            {
                resolveHandler(_result);
            }

            // Rejected
            if (_state == PromiseState.Rejected)
            {
                rejectHandler(_error);
            }
        }

        protected void ResolveBase(TResult result)
        {
            // Sync
            lock (_syncRoot)
            {
                // State
                if (_state != PromiseState.Pending) return;
                _state = PromiseState.Resolved;

                // Results
                _result = result;
                _error = null;
            }

            // Handlers
            List<Action<TResult>> resolveHandlers = _resolveHandlers;
            if (resolveHandlers == null) return;

            // Resolve
            foreach (Action<TResult> resolveHandler in resolveHandlers)
            {
                resolveHandler(result);
            }
        }

        public void Reject(Exception error)
        {
            #region Contracts

            if (error == null) throw new ArgumentNullException();

            #endregion

            // Sync            
            lock (_syncRoot)
            {
                // State
                if (_state != PromiseState.Pending) return;
                _state = PromiseState.Rejected;

                // Result
                _result = default(TResult);
                _error = error;
            }

            // Handlers
            List<Action<Exception>> rejectHandlers = _rejectHandlers;
            if (rejectHandlers == null) return;

            // Reject
            foreach (Action<Exception> rejectHandler in rejectHandlers)
            {
                rejectHandler(error);
            }
        }

        public void Notify(Progress progress)
        {
            #region Contracts

            if (progress == null) throw new ArgumentNullException();

            #endregion

            // Handlers
            List<Action<Progress>> notifyHandlers = null;
            lock (_syncRoot)
            {
                if (_notifyHandlersSnapshot == null && _notifyHandlers != null)
                {
                    _notifyHandlersSnapshot = new List<Action<Progress>>();
                    foreach (Action<Progress> notifyHandler in _notifyHandlers)
                    {
                        _notifyHandlersSnapshot.Add(notifyHandler);
                    }
                }
                notifyHandlers = _notifyHandlersSnapshot;
            }
            if (notifyHandlers == null) return;

            // Notify
            foreach (Action<Progress> notifyHandler in notifyHandlers)
            {
                notifyHandler(progress);
            }
        }
    }
}
